Upstream Diff, for reference only.

PaulLaux opened the pull request:
This PR contains the following modifications that are required for ZSA:
- Add a new `LookupRangeCheck` config named `LookupRangeCheck4_5BConfig` optimized for 4, 5 and 10-bit range checks
- Add variable-base sign-scalar multiplication function called `mul_sign`
- Add mux functionalities into `CondSwapChip`
- Update `SinsemillaChip` to support evaluating a hash from a private point
- Extend `SinsemillaChip` with functions that allow separate evaluation of the hash and blinding factor of a commitment
- Standardize test circuit names with `My...Circuit` convention

daira commented on halo2_gadgets/src/sinsemilla/chip.rs:
```diff
@@ -203,7 +204,7 @@ where
         meta.create_gate("Initial y_Q", |meta| {
             let q_s4 = meta.query_selector(config.q_sinsemilla4);
-            let y_q = meta.query_fixed(config.fixed_y_q);
+            let y_q = meta.query_advice(config.double_and_add.x_p, Rotation::prev());
```
This is incompatible.

daira commented on halo2_gadgets/src/sinsemilla/chip/generator_table.rs:
```diff
@@ -77,6 +78,22 @@ impl GeneratorTableConfig {
         });
     }
 
+    /// Load the generator table into the circuit.
+    ///
+    /// | table_idx |     table_x    |     table_y    | table_range_check_tag |
+    /// -------------------------------------------------------------------
+    /// |     0     |    X(S\[0\])   |    Y(S\[0\])   |           0           |
+    /// |     1     |    X(S\[1\])   |    Y(S\[1\])   |           0           |
+    /// |    ...    |      ...       |       ...      |           0           |
+    /// |   2^10-1  | X(S\[2^10-1\]) | Y(S\[2^10-1\]) |           0           |
+    /// |     0     |    X(S\[0\])   |    Y(S\[0\])   |           4           |
+    /// |     1     |    X(S\[1\])   |    Y(S\[1\])   |           4           |
+    /// |    ...    |       ...      |       ...      |           4           |
+    /// |   2^4-1   | X(S\[2^4-1\])  | Y(S\[2^4-1\])  |           4           |
+    /// |     0     |    X(S\[0\])   |    Y(S\[0\])   |           5           |
+    /// |     1     |    X(S\[1\])   |    Y(S\[1\])   |           5           |
+    /// |    ...    |       ...      |       ...      |           5           |
+    /// |   2^5-1   | X(S\[2^5-1\])  | Y(S\[2^5-1\])  |           5           |
```
Making any change at all to layout is incompatible.

daira commented on halo2_gadgets/src/utilities/lookup_range_check.rs:
```diff
@@ -60,8 +60,11 @@ pub struct LookupRangeCheckConfig<F: PrimeFieldBits, const K: usize> {
     q_lookup: Selector,
     q_running: Selector,
     q_bitshift: Selector,
+    q_range_check_4: Selector,
+    q_range_check_5: Selector,
     running_sum: Column<Advice>,
     table_idx: TableColumn,
+    table_range_check_tag: TableColumn,
```
This is incompatible.

daira commented on halo2_gadgets/src/sinsemilla/chip.rs:
```diff
@@ -178,6 +178,7 @@ where
                 table_idx: lookup.0,
                 table_x: lookup.1,
                 table_y: lookup.2,
+                table_range_check_tag: lookup.3,
```
This is related to the layout range noted above.

YaoJGalteland committed:
OrchardZSA backward compatability using traits (#31)

This commit contains the following modifications:
- Add tests for Lookup, ECC, Merkle, and Sinsemilla to verify that the verification key and the proof have not been modified by comparing them to those saved in a file.
- Introduce a `LookupRangeCheck` trait that provides common methods for a lookup range check.
- Use this new trait as a generic parameter in configs, chips and tests that are using lookup.
- Create a new Lookup chip which is optimized for 4, 5 and 10-bit range check and add tests for this new chip.
- Add `init_from_private_point` parameter in `SinsemillaConfig`. If this parameter is set to false, the `SinsemillaChip` is the same as the current `SinsemillaChip` used in vanilla circuit. Otherwise, the `SinsemillaChip` is modified to support hash from private point.

---------

Co-authored-by: YaoGalteland <yao@qed-it.com>
Co-authored-by: Dmitry Demin <dmitry@qed-it.com>
Co-authored-by: Constance Beguier <constance@qed-it.com>

ConstanceBeguier committed:
Fix clippy warnings

ConstanceBeguier committed:
Fix EccPointQ warnings

PaulLaux committed:
Merge pull request #35 from QED-it/fix_clippy_warnings

Fix clippy warnings

ConstanceBeguier committed:
Merge branch 'main' into merge_main_20250318

ConstanceBeguier committed:
Fix bitrot CI

ConstanceBeguier committed:
Add comments to fix clippy beta

ConstanceBeguier committed:
Merge branch main into zsa1

ConstanceBeguier committed:
Merge branch 'main' into merge_main_20250616

ConstanceBeguier committed:
Merge branch main into zsa1

ConstanceBeguier committed:
Simplify ZSA features (#41)

- Move lookup table `load` function into `LookupRangeCheck` trait (and remove `table_range_check_tag` function which is no longer used)
- Move some imports to reduce the diff with the main branch
- Standardize test circuit names with `My...Circuit` convention
- Reduce visibility of some functions
- Update changelog file

ConstanceBeguier committed:
Add a comment to the rationale for the duplicated-rows range check lookup table (#43)

This PR adds a comment to clarify why a duplicated-rows range check lookup table was chosen.

PaulLaux committed:
Update halo2_gadgets/src/ecc/chip.rs

Co-authored-by: Daira-Emma Hopwood <daira@jacaranda.org>

ConstanceBeguier committed:
Take into account Daira's code review (#44)
